capsules/src/usb.rs

```diff
@@ -0,0 +1,467 @@
+        b[4] = self.class;
+        b[5] = self.subclass;
+        b[6] = self.protocol;
+        b[7] = self.max_packet_size_ep0;
+        put_u16(&mut b[8..10], self.vendor_id);
+        put_u16(&mut b[10..12], self.product_id);
+        put_u16(&mut b[12..14], self.device_release);
+        b[14] = self.manufacturer_string;
+        b[15] = self.product_string;
+        b[16] = self.serial_number_string;
+        b[17] = self.num_configurations;
+        18
+    }
+}
+
+pub struct ConfigurationDescriptor {
+     pub num_interfaces: u8,
+     pub configuration_value: u8,
+     pub string_index: u8,
+     pub attributes: ConfigurationAttributes,
+     pub max_power: u8,   // in 2mA units
+     pub related_descriptor_length: usize,
+}
+
+impl Default for ConfigurationDescriptor {
+    fn default() -> Self {
+        ConfigurationDescriptor {
+            num_interfaces: 1,
+            configuration_value: 0,
+            string_index: 0,
+            attributes: ConfigurationAttributes::new(true, false),
+            max_power: 0,   // in 2mA units
+            related_descriptor_length: 0
+        }
+    }
+}
+
+impl Descriptor for ConfigurationDescriptor {
+    fn size(&self) -> usize { 9 }
+
+    fn write_to(&self, buf: &mut [u8]) -> usize {
+        buf[0] = 9; // Size of descriptor
+        buf[1] = DescriptorType::Configuration as u8;
+        put_u16(&mut buf[2..4], (9 + self.related_descriptor_length) as u16);
+        buf[4] = self.num_interfaces;
+        buf[5] = self.configuration_value;
+        buf[6] = self.string_index;
+        buf[7] = From::from(self.attributes);
+        buf[8] = self.max_power;
+        9
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct ConfigurationAttributes(u8);
+
+impl ConfigurationAttributes {
+    pub fn new(is_self_powered: bool, supports_remote_wakeup: bool) -> Self {
+        ConfigurationAttributes((1 << 7)
+                                | if is_self_powered { 1 << 6 } else { 0 }
+                                | if supports_remote_wakeup { 1 << 5 } else { 0 })
+    }
+}
+
+impl From<ConfigurationAttributes> for u8 {
+    fn from(ca: ConfigurationAttributes) -> u8 {
+        ca.0
+    }
+}
+
+pub struct InterfaceDescriptor {
+    pub interface_number: u8,
+    pub alternate_setting: u8,
+    pub num_endpoints: u8,
+    pub interface_class: u8,
+    pub interface_subclass: u8,
+    pub interface_protocol: u8,
+    pub string_index: u8,
+}
+
+impl Default for InterfaceDescriptor {
+    fn default() -> Self {
+        InterfaceDescriptor {
+            interface_number: 0,
+            alternate_setting: 0,
+            num_endpoints: 0, // (exluding default control endpoint)
+            interface_class: 0xff, // vendor_specific
+            interface_subclass: 0xab,
+            interface_protocol: 0,
+            string_index: 0,
+        }
+    }
+}
+
+impl Descriptor for InterfaceDescriptor {
+    fn size(&self) -> usize { 9 }
+
+    fn write_to(&self, buf: &mut [u8]) -> usize {
```

Comment:
should bounds-check here. since it's only 9 bytes, it might be simplest to create the buffer on the stack with the expected size and use a fold or a built in copy function or something (might even get inlined out... who knows...)